python/pyspark/sql/context.py

```diff
@@ -336,19 +353,28 @@ def createDataFrame(self, rdd, schema=None, samplingRatio=None):
         >>> df3.collect()
         [Row(name=u'Alice', age=1)]
         """
-        if isinstance(rdd, DataFrame):
+        if isinstance(data, DataFrame):
             raise TypeError("rdd is already a DataFrame")
 
-        if isinstance(schema, StructType):
-            return self.applySchema(rdd, schema)
-        else:
-            if isinstance(schema, (list, tuple)):
-                first = rdd.first()
-                if not isinstance(first, (list, tuple)):
-                    raise ValueError("each row in `rdd` should be list or tuple")
-                row_cls = Row(*schema)
-                rdd = rdd.map(lambda r: row_cls(*r))
-            return self.inferSchema(rdd, samplingRatio)
+        if has_pandas and isinstance(data, pandas.DataFrame):
+            data = self._sc.parallelize(data.to_records(index=False))
+            if schema is None:
+                schema = list(data.columns)
+
+        if not isinstance(data, RDD):
+            data = self._sc.parallelize(data)
```

Comment:
can we provide more type checking? e.g. throw a TypeError if data is not the proper type. maybe do that check in the very beginning.